server/listen.ts

```diff
@@ -0,0 +1,153 @@
+
+      lastBlockNumberBeingProcessed = blockNumber + 1
+
+      const [offerCreatedFilter, offerAcceptedFilter] = await Promise.all([
+        publicClient.createContractEventFilter({
+          abi: parseAbi(['event OfferCreated(string indexed domainName, address indexed srcAsset, address indexed destAsset, address offerAddress, address domainOwner, uint256 depositAmount, uint256 closeAmount, uint256 commissionRate, uint256 lockWithdrawAfter)']),
+          address: String(process.env.OTC_CONTRACT) as Address,
+          fromBlock: BigInt(lastBlockNumber),
+          toBlock: BigInt(blockNumber),
+          eventName: 'OfferCreated'
+        }),
+
+        publicClient.createContractEventFilter({
+          abi: OFFER_ABI.abi,
+          fromBlock: BigInt(lastBlockNumber),
+          toBlock: BigInt(blockNumber),
+          eventName: 'OfferAccepted'
+        })
+      ])
+
+      const [createdLogs, acceptedLogs] = await Promise.all([
+        publicClient.getFilterLogs({ filter: offerCreatedFilter }),
+        publicClient.getFilterLogs({ filter: offerAcceptedFilter })
+      ])
+
+      for (const log of createdLogs) {
+        if (log.eventName === 'OfferCreated') {
+          const [srcPrice, destPrice, time] = await Promise.all([
+            getPrice(log.args.srcAsset!),
+            getPrice(log.args.destAsset!),
+            getTimestamp(Number(log.blockNumber))
+          ])
+
+          await pool.query(`
+            INSERT INTO
+              logs(event_name, time, src_asset, dest_asset, offer_address, domain_owner, close_amount, total_deposits, src_price, dest_price)
+            VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
+          `, [
+            'OfferCreated',
+            new Date(time).toUTCString(),
+            String(log.args.srcAsset).toLowerCase(),
+            String(log.args.destAsset).toLowerCase(),
+            String(log.args.offerAddress).toLowerCase(),
+            String(log.args.domainOwner).toLowerCase(),
+            log.args.closeAmount,
+            log.args.depositAmount,
+            srcPrice,
+            destPrice,
+          ])
+        }
+      }
+
+      for (const log of acceptedLogs) {
+        const [srcAsset, destAsset, domainOwner, closeAmount, totalDeposits] = await Promise.all(
+          ['srcAsset', 'destAsset', 'domainOwner', 'acceptAmount', 'total_deposits']
+            .map((func) => publicClient.readContract({
+              address: log.address,
+              abi: OFFER_ABI.abi,
+              functionName: func
+            }))
+        )
+
+        const [srcPrice, destPrice, time] = await Promise.all([
+          getPrice(String(srcAsset)),
+          getPrice(String(destAsset)),
+          getTimestamp(Number(log.blockNumber))
+        ])
+
+        await pool.query(`
+          INSERT INTO
+            logs(event_name, time, src_asset, dest_asset, offer_address, domain_owner, close_amount, total_deposits, src_price, dest_price)
```

Comment:
Not guarding against duplicates. Should create deterministic id based on data (transaction hash, event position in transaction, ...)

Reply:
What is the duplication?

Reply:
How about adding unique constraint to table?
```
  UNIQUE (event_name, offer_address)
```

Reply:
Okay, make it fail-safe - should simply log (to file or console) constraint violation and continue with next record

Reply:
Done